File: validation/validation-annotation-processor/src/test/java/io/koraframework/validation/annotation/processor/ValidationExtensionTest.java
package io.koraframework.validation.annotation.processor;

import org.junit.jupiter.api.Test;
import io.koraframework.kora.app.annotation.processor.KoraAppProcessor;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class ValidationExtensionTest extends AbstractValidationAnnotationProcessorTest {

    @Test
    public void testExtension() {
        var compileResult = compile(List.of(new KoraAppProcessor(), new ValidAnnotationProcessor()),
            """
                @Valid
                public record TestRecord(@Size(min = 1, max = 5) java.util.List<String> list){}
                """, """
                @KoraApp
                public interface TestApp extends ValidatorModule{
                   @Root
                   default String root(Validator<TestRecord> testRecordValidator) { return "";}
                }
                """);
        compileResult.assertSuccess();

        var validatorClass = compileResult.loadClass("$TestRecord_Validator");
        assertThat(validatorClass).isNotNull();
        var graph = compileResult.loadClass("TestAppGraph");
        assertThat(graph).isNotNull();
    }
}

File: validation/validation-symbol-processor/src/test/kotlin/io/koraframework/validation/symbol/processor/ValidationExtensionTest.kt
package io.koraframework.validation.symbol.processor

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import io.koraframework.kora.app.ksp.KoraAppProcessorProvider
import io.koraframework.ksp.common.AbstractSymbolProcessorTest

class ValidationExtensionTest : AbstractSymbolProcessorTest() {
    @Test
    fun testExtension() {
        compile0(listOf(KoraAppProcessorProvider(), ValidSymbolProcessorProvider()),
            """
                import io.koraframework.validation.common.annotation.Size
                import io.koraframework.validation.common.annotation.Valid

                @Valid
                data class TestRecord(@Size(min = 1, max = 5) val list: List<String>) {}
                
                """.trimIndent(),
            """
                import io.koraframework.common.KoraApp;
                import io.koraframework.common.annotation.Root;
                import io.koraframework.validation.common.Validator;
                import io.koraframework.validation.common.constraint.ValidatorModule;
                @KoraApp
                interface TestApp : ValidatorModule {
                   @Root
                   fun root(testRecordValidator: Validator<TestRecord>) = ""
                }
                
                """.trimIndent()
        )
        compileResult.assertSuccess()
        val validatorClass = loadClass("\$TestRecord_Validator")
        assertThat(validatorClass).isNotNull()
        val graph = loadClass("TestAppGraph")
        assertThat(graph).isNotNull()
    }
}
